### ISSUE TITLE
Don't pass keys by value

### ISSUE THREAD

**Issue by jamii** [tigerbeetle/tigerbeetle#566]
A lot of hotspots happen to be places where we also pass keys by value instead of by reference. There may be some cheap wins here.

* Change [key_from_value](https://github.com/tigerbeetledb/tigerbeetle/blob/28373c2702f555d8db507e104b221e9b34f7cc60//src/lsm/table.zig#L77) to return `*const Key`, and change `compare_keys` to take `*const Key`.
* Change [binary_search](https://github.com/tigerbeetledb/tigerbeetle/blob/28373c2702f555d8db507e104b221e9b34f7cc60//src/lsm/binary_search.zig#L22) to use `*const Key`.
* Vendor std.sort and use `*const Key`.
* Vendor std.HashMap and use `*const Key`.
* Vendor std.crypto.blake3 and pass arrays by reference instead of by value.

**Comment by jorangreef** [comment]
I'll fall off my chair if this is the case (and I have a feeling it is)... 🪑

**Comment by jorangreef** [comment]
@batiati would you be able to tackle this?

The motivation is that we're concerned about stack copies impacting storage engine performance.

**Comment by jorangreef** [comment]
We also want to be more explicit about our intention to the compiler (and don't want to say "trust the compiler") because otherwise we risk performance regressions down the road if the compiler is ever changed. And here, the behavior of pass-by-value in Zig is subject to change if I recall correctly.

**Comment by batiati** [comment]
Not related to passing keys by ref, ~but I found this~:

On [lsm/composite_keys.zig](https://github.com/tigerbeetledb/tigerbeetle/blob/28373c2702f555d8db507e104b221e9b34f7cc60/src/lsm/composite_key.zig#L45-L57):

```diff
        pub inline fn compare_keys(a: Self, b: Self) math.Order {
-            if (a.field < b.field) {
-                return .lt;
-            } else if (a.field > b.field) {
-                return .gt;
-            } else if (a.timestamp < b.timestamp) {
-                return .lt;
-            } else if (a.timestamp > b.timestamp) {
-                return .gt;
-            } else {
-                return .eq;
-            }
+            const Int = std.meta.Int(.unsigned, @bitSizeOf(Self));
+            return math.order(@bitCast(Int, a), @bitCast(Int, b));
        }
```

Before:

```
1223 batches in 136.85 s
load offered = 1000000 tx/s
load accepted = 73070 tx/s
batch latency p00 = 5 ms
batch latency p10 = 37 ms
batch latency p20 = 42 ms
batch latency p30 = 42 ms
batch latency p40 = 42 ms
batch latency p50 = 43 ms
batch latency p60 = 47 ms
batch latency p70 = 47 ms
batch latency p80 = 52 ms
batch latency p90 = 58 ms
batch latency p100 = 7037 ms
transfer latency p00 = 5 ms
transfer latency p10 = 6852 ms
transfer latency p20 = 14791 ms
transfer latency p30 = 23691 ms
transfer latency p40 = 37023 ms
transfer latency p50 = 49655 ms
transfer latency p60 = 63697 ms
transfer latency p70 = 79547 ms
transfer latency p80 = 96578 ms
transfer latency p90 = 111073 ms
transfer latency p100 = 126773 ms
```

After:

```
1223 batches in 104.57 s
load offered = 1000000 tx/s
load accepted = 95629 tx/s
batch latency p00 = 5 ms
batch latency p10 = 42 ms
batch latency p20 = 42 ms
batch latency p30 = 42 ms
batch latency p40 = 47 ms
batch latency p50 = 47 ms
batch latency p60 = 47 ms
batch latency p70 = 47 ms
batch latency p80 = 52 ms
batch latency p90 = 64 ms
batch latency p100 = 4714 ms
transfer latency p00 = 5 ms
transfer latency p10 = 4149 ms
transfer latency p20 = 14925 ms
transfer latency p30 = 24823 ms
transfer latency p40 = 33314 ms
transfer latency p50 = 42946 ms
transfer latency p60 = 52237 ms
transfer latency p70 = 61992 ms
transfer latency p80 = 71598 ms
transfer latency p90 = 84074 ms
transfer latency p100 = 94541 ms
```

**EDITED**

Not sure if I messed with the comparison when I did `@bitCast`. It runs faster, but maybe it's just doing less work ... I noticed the data file grows only 6GiB with this change, while running the original code, it grows up to 14GiB.
Probably it's a false-positive finding, and I need to dig more to be sure.

**EDITED 2**

Yes, I messed it up, a lesser `field` should return `.lt` even with a greater `timestamp` ... the correct logic should bitcast using `field` as the MSB, but there isn't any performance gain by doing that 🥲

**Comment by batiati** [comment]
After spending some time on experiments, here is a summary of my findings on this issue so far:

TL;DR;
None of these actions had any positive impact on benchmark results, and some of them have made the code harder to read/maintain.

### Params by ref

- Change `compare_keys` to take `*const Key`.
- Change binary_search to use `*const Key`.

Since those functions are generic, in many places, the real type will be a `timestamp: u64` or an `id: u128`.
I decided to try passing by ref everywhere, and if we could notice some improvement, I'd bother to create something more sophisticated to avoid passing `u64` by ref.

Experimental changes can be found [here](https://github.com/tigerbeetledb/tigerbeetle/compare/main...batiati-by-ref)

No performance improvement (just noise):

Before:
```
Benchmarking...
1223 batches in 144.02 s
load offered = 1000000 tx/s
load accepted = 69436 tx/s
batch latency p00 = 5 ms
batch latency p10 = 42 ms
batch latency p20 = 47 ms
batch latency p30 = 47 ms
batch latency p40 = 47 ms
batch latency p50 = 47 ms
batch latency p60 = 48 ms
batch latency p70 = 48 ms
batch latency p80 = 53 ms
batch latency p90 = 58 ms
batch latency p100 = 7046 ms
transfer latency p00 = 5 ms
transfer latency p10 = 4760 ms
transfer latency p20 = 12839 ms
transfer latency p30 = 23419 ms
transfer latency p40 = 35741 ms
transfer latency p50 = 50482 ms
transfer latency p60 = 65840 ms
transfer latency p70 = 83596 ms
transfer latency p80 = 99281 ms
transfer latency p90 = 117303 ms
transfer latency p100 = 133981 ms
```

After:
```
1223 batches in 141.50 s
load offered = 1000000 tx/s
load accepted = 70673 tx/s
batch latency p00 = 5 ms
batch latency p10 = 42 ms
batch latency p20 = 43 ms
batch latency p30 = 47 ms
batch latency p40 = 47 ms
batch latency p50 = 47 ms
batch latency p60 = 48 ms
batch latency p70 = 48 ms
batch latency p80 = 48 ms
batch latency p90 = 62 ms
batch latency p100 = 6956 ms
transfer latency p00 = 5 ms
transfer latency p10 = 5065 ms
transfer latency p20 = 13353 ms
transfer latency p30 = 23440 ms
transfer latency p40 = 36215 ms
transfer latency p50 = 52348 ms
transfer latency p60 = 67248 ms
transfer latency p70 = 83698 ms
transfer latency p80 = 98680 ms
transfer latency p90 = 115362 ms
transfer latency p100 = 131447 ms
```

### Return by ref

- Change `key_from_value` to return `*const Key`.

Not all implementations of `key_from_value` can return a reference, sometimes it returns a stack value, like [lsm/composite_key.zig#L59](https://github.com/tigerbeetledb/tigerbeetle/blob/088361ef8ea176c5b0d1d86a075af8c56a231813/src/lsm/composite_key.zig#L59-L64) and [groove.zig#L27](https://github.com/tigerbeetledb/tigerbeetle/blob/088361ef8ea176c5b0d1d86a075af8c56a231813/src/lsm/groove.zig#L27-L29):

```zig
pub inline fn key_from_value(value: *const Value) Self {
    return .{
        .field = value.field,
        .timestamp = @truncate(u63, value.timestamp),
    };
}
```

I tried to make it easy for the call sites to deal with both signatures, but this change would be too complicated. 
Besides that, inline functions are supposed to behave like macros, no matter if the parameter/return is by value, they shouldn't ever be copied, right? 🤔

I can explore it further if we decide to go for that.

### Sort

Replaced occurrences of `std.sort` with a [custom version](https://github.com/batiati/zig/commit/d2c892c784ba7ec5436c9c76bcc543361caead21). which `compareFn` takes `*const T`.

Experimental changes can be found [here](https://github.com/tigerbeetledb/tigerbeetle/compare/main...batiati-sort)

No noticeable improvement (ran three times each, picked the best one):

Before
```
Benchmarking...
1223 batches in 131.13 s
load offered = 1000000 tx/s
load accepted = 76258 tx/s
batch latency p00 = 5 ms
batch latency p10 = 37 ms
batch latency p20 = 37 ms
batch latency p30 = 42 ms
batch latency p40 = 42 ms
batch latency p50 = 42 ms
batch latency p60 = 42 ms
batch latency p70 = 47 ms
batch latency p80 = 47 ms
batch latency p90 = 57 ms
batch latency p100 = 7278 ms
transfer latency p00 = 5 ms
transfer latency p10 = 5647 ms
transfer latency p20 = 13826 ms
transfer latency p30 = 22465 ms
transfer latency p40 = 32942 ms
transfer latency p50 = 47475 ms
transfer latency p60 = 61420 ms
transfer latency p70 = 75043 ms
transfer latency p80 = 89236 ms
transfer latency p90 = 105700 ms
transfer latency p100 = 121105 ms
```

After
```
Benchmarking...
1223 batches in 131.55 s
load offered = 1000000 tx/s
load accepted = 76013 tx/s
batch latency p00 = 5 ms
batch latency p10 = 37 ms
batch latency p20 = 37 ms
batch latency p30 = 42 ms
batch latency p40 = 42 ms
batch latency p50 = 42 ms
batch latency p60 = 42 ms
batch latency p70 = 47 ms
batch latency p80 = 48 ms
batch latency p90 = 57 ms
batch latency p100 = 6440 ms
transfer latency p00 = 5 ms
transfer latency p10 = 3875 ms
transfer latency p20 = 10559 ms
transfer latency p30 = 20829 ms
transfer latency p40 = 31954 ms
transfer latency p50 = 45051 ms
transfer latency p60 = 58974 ms
transfer latency p70 = 72893 ms
transfer latency p80 = 87502 ms
transfer latency p90 = 104994 ms
transfer latency p100 = 121528 ms
```

### Hashmaps

The custom HashMap supporting `hash` and `eql` by ref can be found [here](https://github.com/batiati/zig/commit/3d89e169940492bdb33b85c7bc7c648e0186363c)

These are the points I felt most confident we could have some improvement:

[lsm/groove.zig](https://github.com/tigerbeetledb/tigerbeetle/blob/75f85b6bf3ca269e3c7265e8d19c91f9e8672d7c/src/lsm/groove.zig#L413-L430)

```diff zig
        const PrefetchObjectsContext = struct {
-           pub fn hash(_: PrefetchObjectsContext, object: Object) u64 {
+           pub fn hash(_: PrefetchObjectsContext, object: *const Object) u64 {
                return std.hash.Wyhash.hash(0, mem.asBytes(&@field(object, primary_field)));
            }

-           pub fn eql(_: PrefetchObjectsContext, a: Object, b: Object) bool {
+           pub fn eql(_: PrefetchObjectsContext, a: *const Object, b: *const Object) bool {
                return @field(a, primary_field) == @field(b, primary_field);
            }
        };
        const PrefetchObjectsAdapter = struct {
            pub fn hash(_: PrefetchObjectsAdapter, key: PrimaryKey) u64 {
                return std.hash.Wyhash.hash(0, mem.asBytes(&key));
            }

-           pub fn eql(_: PrefetchObjectsAdapter, a_key: PrimaryKey, b_object: Object) bool {
-               return a_key == @field(b_object, primary_field);
+           pub fn eql(_: PrefetchObjectsAdapter, a_key: *const PrimaryKey, b_object: *const Object) bool {
+               return a_key.* == @field(b_object, primary_field);
            }
        };

```

No noticeable improvement (ran three times each, picked the best one):

Before:
```
Benchmarking...
1223 batches in 133.04 s
load offered = 1000000 tx/s
load accepted = 75166 tx/s
batch latency p00 = 5 ms
batch latency p10 = 37 ms
batch latency p20 = 42 ms
batch latency p30 = 42 ms
batch latency p40 = 42 ms
batch latency p50 = 42 ms
batch latency p60 = 42 ms
batch latency p70 = 47 ms
batch latency p80 = 52 ms
batch latency p90 = 64 ms
batch latency p100 = 5937 ms
transfer latency p00 = 5 ms
transfer latency p10 = 4809 ms
transfer latency p20 = 14680 ms
transfer latency p30 = 23807 ms
transfer latency p40 = 36226 ms
transfer latency p50 = 49950 ms
transfer latency p60 = 65156 ms
transfer latency p70 = 79184 ms
transfer latency p80 = 93301 ms
transfer latency p90 = 107891 ms
transfer latency p100 = 123011 ms
```

After:
```
Stopping replica 0...

Benchmarking...
1223 batches in 136.20 s
load offered = 1000000 tx/s
load accepted = 73422 tx/s
batch latency p00 = 5 ms
batch latency p10 = 37 ms
batch latency p20 = 42 ms
batch latency p30 = 42 ms
batch latency p40 = 42 ms
batch latency p50 = 43 ms
batch latency p60 = 47 ms
batch latency p70 = 47 ms
batch latency p80 = 48 ms
batch latency p90 = 57 ms
batch latency p100 = 6037 ms
transfer latency p00 = 5 ms
transfer latency p10 = 4518 ms
transfer latency p20 = 13606 ms
transfer latency p30 = 22616 ms
transfer latency p40 = 35068 ms
transfer latency p50 = 49067 ms
transfer latency p60 = 62934 ms
transfer latency p70 = 79341 ms
transfer latency p80 = 95126 ms
transfer latency p90 = 110822 ms
transfer latency p100 = 126168 ms
```


### Blake 3

Passing arrays by ref have no noticiable effect on performance, also I updated `blake3.zig` to the latest master version and replaced all `std.mem.copy` for `@memcpy`. Looks like Zig compiler does a nice job here.

[Here's the diff](https://github.com/batiati/zig/commit/851dff14e48032253a3f400596874da6cdf527ef).

Benchmark Zig 0.9.1 ReleaseSafe on i7-1185G7 @ 3.00GHz, executed 3 times:

```bash
zig run -O ReleaseSafe --zig-lib-dir ./lib ./lib/std/crypto/benchmark.zig
```

Before:
```
blake3:        986 MiB/s
blake3:        871 MiB/s
blake3:        977 MiB/s
```

After:
```
blake3:        891 MiB/s
blake3:        974 MiB/s
blake3:        969 MiB/s
```

-------------------

Any suggestions for the next steps? I'll be happy to try other strategies.

**Comment by jamii** [comment]
> Any suggestions for the next steps?

I got nothing :S 

Thanks for the thorough experiments.
